p2p/server.py

```diff
@@ -0,0 +1,136 @@
+import asyncio
+import logging
+import secrets
+
+from typing import Tuple
+from eth_keys import datatypes
+
+from evm.db.chain import AsyncChainDB
+from p2p.auth import (
+    decode_auth_plain,
+    decode_auth_eip8,
+    HandshakeResponder,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.discovery import DiscoveryProtocol
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import OperationCancelled
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import (
+    ETHPeer,
+    PeerPool,
+)
+from p2p.utils import sxor
+
+
+class Server:
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+
+    def __init__(self,
+                 privkey: datatypes.PrivateKey,
+                 server_address: Address,
+                 chaindb: AsyncChainDB
+                 ) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.chaindb = chaindb
+        self.privkey = privkey
+        self.server_address = server_address
+        discovery = DiscoveryProtocol(self.privkey, self.server_address, bootstrap_nodes=[])
+        self.peer_pool = PeerPool(ETHPeer, self.chaindb, 1, self.privkey, discovery)
+
+    async def run(self) -> None:
+        self.logger.info("Running server...")
+        loop = asyncio.get_event_loop()
+        factory = asyncio.start_server(
+            self.receive_handshake, host=self.server_address.ip, port=self.server_address.udp_port)
+        asyncio.ensure_future(factory)
+
+        while not self.cancel_token.triggered:
+            try:
+                loop.run_forever()
+            except OperationCancelled:
+                break
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        asyncio.ensure_future(self.peer_pool.stop())
+
+    async def receive_handshake(
+            self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        # Use reader to read the auth_init msg until EOF
+        msg = await reader.read()
+
+        # Use HandshakeResponder.decode_authentication(auth_init_message) on auth init msg
+        ephem_pubkey, initiator_nonce, initiator_pubkey = decode_authentication(msg, self.privkey)
+
+        # Get remote's address: IPv4 or IPv6
+        peername = writer.get_extra_info("peername")
+        ip = peername[0]
+        port = peername[1]
```

Comment:
Can you use sequence unpacking here?

Reply:
If the address is IPv4 it returns 2 elements, but if it's IPv6 it returns 4 elements, which is why I went with this setup. Is it safe to assume only IPv4 addresses?

Reply:
@njgheorghita you can do this in python 3

```python
In [7]: a, b, *c = [1, 2, 3, 4]

In [8]: print(a, b, c)
1 2 [3, 4]

In [9]: a, b, *c = [1, 2]

In [10]: print(a, b, c)
1 2 []
```

So you can just add an exta `*_` to the unpacking to catch the extra that you don't plan to use (`_`  is the convention for assigning to a variable that you don't plan to use).

```python
ip, port, *_ = peername
```